hail/src/main/scala/is/hail/expr/types/physical/PCanonicalBinary.scala

```diff
@@ -105,6 +83,61 @@ class PCanonicalBinary(val required: Boolean) extends PBinary {
       Region.storeInt(addr, bytes.length),
       Region.storeBytes(bytesOffset(addr), bytes)
     )
+
+  def constructAtAddress(mb: MethodBuilder, addr: Code[Long], region: Code[Region], srcPType: PType, srcAddress: Code[Long], forceDeep: Boolean): Code[Unit] = {
+    val srcBinary = srcPType.asInstanceOf[PBinary]
+
+    val srcAddrVar = mb.newLocal[Long]
+    val len = mb.newLocal[Int]
+    Code(
+      srcAddrVar := srcAddress,
+      len := srcBinary.loadLength(srcAddrVar),
+      Region.storeAddress(addr, constructOrCopyWithLen(mb, region, srcBinary, srcAddrVar, len, forceDeep))
+    )
+  }
+
+  def constructOrCopyWithLen(mb: MethodBuilder, region: Code[Region], srcBinary: PBinary, srcAddress: LocalRef[Long], len: LocalRef[Int], forceDeep: Boolean): Code[Long] = {
+    if (srcBinary == this) {
+      if (forceDeep) {
+        val newAddr = mb.newLocal[Long]
+        Code(
+          newAddr := allocate(region, len),
+          Region.copyFrom(srcAddress, newAddr, contentByteSize(len)),
+          newAddr)
+      } else
+        srcAddress
```

Comment:
I prefer putting the simpler case first, and using early return, because I think complex if-else blocks are harder to read. I'm not clear what our style is, but if we settle on one style, I think we should add that to the style guide.

Reply:
I think early return certainly has its uses, but don't see it making things more readable here. Certainly up to personal preference.